src/FsAutoComplete.Core/Commands.fs

```diff
@@ -26,54 +27,67 @@ type Commands (serialize : Serializer) =
     member private x.SerializeResult (successToString: Serializer -> 'a -> string, ?failureToString: Serializer -> string -> string) =
         x.SerializeResultAsync ((fun s x -> successToString s x |> async.Return), ?failureToString = failureToString)
 
-    member __.TryGetRecentTypeCheckResultsForFile = checker.TryGetRecentCheckResultsForFile
-    member __.TryGetFileCheckerOptionsWithLinesAndLineStr = state.TryGetFileCheckerOptionsWithLinesAndLineStr
-    member __.TryGetFileCheckerOptionsWithLines = state.TryGetFileCheckerOptionsWithLines
-    member __.Files = state.Files
+    member private x.AsCancellable (filename : SourceFilePath) (action : Async<string list>) =
+        let cts = new CancellationTokenSource()
+        state.AddCancellationToken(filename, cts)
+        Async.StartCatchCancellation(action, cts.Token)
+        |> Async.Catch
+        |> Async.map (function Choice1Of2 res -> res | Choice2Of2 _ -> [Response.info serialize "Request cancelled"])
```

Comment:
~~Sadly cancellation in Async is quite of frustrating to deal with.
I don't think you actually get `Choice2Of2` in the state of cancellation as the `Async` completely "stops" (see https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/FSharp.Core/control.fs#L1041)~~

Ok nevermind the StartCatchCancellation fixes the problem

Reply:
Also it is always a bad idea to "throw error information" away like this. At the very least it should be logged.

Reply:
It's not really error in this case, as far as I understand it will happen only if we cancel request. The "real" error handling is done in lower layer inside commands logic.